### PR TITLE
InputStream to read from S3

### PR DESCRIPTION
**Pull request by ivankelly**
Reads from S3 in configurable chunks. Interface allows caller to seek
to an arbitrary position.

Master Issue: #1511

### CODE REVIEW

**Comment by ivankelly** [comment]
retest this please //  V1_ReplicatorTest.testCloseReplicatorStartProducer

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    private final int bufferSize;
+
+    private long cursor;
+
+    public S3BackedInputStreamImpl(AmazonS3 s3client, String bucket, String key,
+                                   long objectLen, int bufferSize) {
+        this.s3client = s3client;
+        this.bucket = bucket;
+        this.key = key;
+        this.buffer = PooledByteBufAllocator.DEFAULT.heapBuffer(bufferSize, bufferSize);
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java]: any specific reason you are using `heapBuffer` here? why can't you use `buffer`?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            log.debug("Reading range {}-{} from {}/{}", startRange, endRange, bucket, key);
+            try (S3Object obj = s3client.getObject(req)) {
+                Long[] range = obj.getObjectMetadata().getContentRange();
+                long bytesRead = range[1] - range[0] + 1;
+
+                buffer.clear();
+                InputStream s = obj.getObjectContent();
+                long bytesToCopy = bytesRead;
+                while (bytesToCopy > 0) {
+                    bytesToCopy -= buffer.writeBytes(s, (int)bytesToCopy);
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java]: don't you need to reset `writerIndex` and `readerIndex` before copying the data?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        this.bucket = bucket;
+        this.key = key;
+        this.buffer = PooledByteBufAllocator.DEFAULT.heapBuffer(bufferSize, bufferSize);
+        this.objectLen = objectLen;
+        this.bufferSize = bufferSize;
+        this.cursor = 0;
+    }
+
+    @Override
+    public int read() throws IOException {
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java]: It is probably better to override `int read(byte b[], int off, int len)`?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            GetObjectRequest req = new GetObjectRequest(bucket, key)
+                .withRange(startRange, endRange);
+            log.debug("Reading range {}-{} from {}/{}", startRange, endRange, bucket, key);
+            try (S3Object obj = s3client.getObject(req)) {
+                Long[] range = obj.getObjectMetadata().getContentRange();
+                long bytesRead = range[1] - range[0] + 1;
+
+                buffer.clear();
+                InputStream s = obj.getObjectContent();
+                long bytesToCopy = bytesRead;
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java]: why not cast it to int here?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java (hunk 1 above)
Comment (ivankelly) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java]: copy paste. changed to buffer

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java (hunk 2 above)
Comment (ivankelly) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java]: I clear on L75.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java (hunk 3 above)
Comment (ivankelly) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java]: Added an override

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java (hunk 4 above)
Comment (ivankelly) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/S3BackedInputStreamImpl.java]: changed

**Comment by zhaijack** [comment]
retest this please

**Pull request merged**